keps/sig-network/20190603-endpointslices/README.md

```diff
@@ -626,6 +678,173 @@ address types.
 3. Deprecate the `IP` address type, making it invalid for new EndpointSlices in
 1.17 before becoming fully invalid in 1.18.
 
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - *EndpointSlice*: Guards controller and API server enablement.
+    - *EndpointSliceProxying*: Guards kube-proxy enablement
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  Reenabling the `EndpointSlice` feature gate could result in the controller
+  cleaning up or updating stale EndpointSlices. Reenabling the
+  `EndpointSliceProxying` feature gate should not result in any changes to API
+  resources.
+
+* **Are there any tests for feature enablement/disablement?**
+  There are unit tests covering kube-proxy with and without the EndpointSlice
+  functionality enabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  The biggest risk comes from enabling EndpointSlice functionality as part of an
+  upgrade. The controller needs to be enabled before kube-proxy to ensure it has
+  sufficient time to create EndpointSlices for all endpoints in the cluster. If
+  kube-proxy is started before then and configured to read from EndpointSlices,
+  only the endpoints that have been represented with EndpointSlices at that
+  point in time will be accessible from that Node. 
+
+* **What specific metrics should inform a rollback?**
+
+* **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
+  This has been manually tested on a variety of e2e clusters by manually
+  enabling or disabling the feature gate and/or API for api-server,
+  controller-manager, and kube-proxy.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features,
+  APIs, fields of API types, flags, etc.?**
+  No.
+
+### Monitoring requirements
+
+* **How can an operator determine if the feature is in use by workloads?**
+  There are a variety of metrics that indicate the quantity and efficiency of
+  EndpointSlices in the cluster. These include:
+  - endpoint_slice_controller/endpoints_added_per_sync
+  - endpoint_slice_controller/endpoints_removed_per_sync
+  - endpoint_slice_controller/endpoints_desired
+	- endpoint_slice_controller/num_endpoint_slices
+	- endpoint_slice_controller/desired_endpoint_slices
+	- endpoint_slice_controller/changes
+
+* **What are the SLIs (Service Level Indicators) an operator can use to
+  determine the health of the service?**
+  - [x] Metrics
+    - Metric name: network_programming_duration_seconds
```

Comment:
You may want to explicitly link to: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/network_programming_latency.md

Reply:
Good call, done.